### ISSUE TITLE
View event causes the vscode extension to be unresponsive

### ISSUE THREAD

**Issue by lachrist** [applandinc/appmap-js#335]
Hi team!

When I click on the "View event in trace view" eye button, the vscode extension becomes unresponsive. More precisely, I still see animations such as button highligthing but nothing happens when I click on them. Only the zoom range seems to still produce an action. The same thing happens when I click on the "Show in trace" button. 

<img width="1392" alt="Screenshot 2021-09-04 at 08 22 34" src="https://user-images.githubusercontent.com/5468597/132085247-b848a935-16bf-4a67-ba06-f5bc06cb2786.png">

* file: https://github.com/land-of-apps/juice-setup/blob/2232503e46b2fe7a4912035086cc48b84b6a7c6d/juice1.appmap.json
* vscode version: 1.60.0
* vscode appmap extension version: v0.17.2.

**Comment by sashahonchar** [comment]
@lachrist I suppose webview becomes unresponsive because of huge size of appmap file.
I tried to open provided appmap in Storybook, but I got an error:
![image](https://user-images.githubusercontent.com/1647695/132251586-d0727f07-422f-4f1b-bdf5-91f575ea4233.png)

I'm going to investigate what's wrong with that appmap because I need an ability to reproduce this issue.

**Comment by lachrist** [comment]
Hi @sashahonchar , thanks for the reply. This issue persists with build v0.18.0... I further investigated the error. The event #250 is in fact returned by event #251. The indexes are not continuous nor sorted, could that be a problem? I assumed they have no meaning but identifying call/return pairs.

```json
{
      "event": "call",
      "thread_id": 0,
      "id": 250,
      "defined_class": "servePrivacyPolicyProof",
      "method_id": "()",
      "path": "build/routes/privacyPolicyProof.js",
      "lineno": 10,
      "static": false,
      "receiver": {
        "name": "this",
        "class": "undefined",
        "object_id": null,
        "value": "undefined"
      },
      "parameters": []
}
```

```json
{
      "id": 251,
      "event": "return",
      "thread_id": 0,
      "parent_id": 250,
      "elapsed": 2.1885160207748413
}
```

And I used `@appland/appmap-validate` to make sure the stack was not messed up. The stack is indeed pretty big but is well formed. Checkout the stack file for a visualization.

Appmap: https://raw.githubusercontent.com/applandinc/appmap-js/70fcaed1e2c04b2edbc696e03f018b8c8a76189f/packages/validate/example/1-6-0/juice.appmap.json
Stack: https://raw.githubusercontent.com/applandinc/appmap-js/70fcaed1e2c04b2edbc696e03f018b8c8a76189f/packages/validate/example/1-6-0/stack.txt

**Comment by kgilpin** [comment]
So, something you need to know about how the AppMaps are loaded - the
loader goes through all the events and re-numbers them sequentially from 1,
if you call "normalize".

So the event id 1423 is probably not 1423 in the AppMap file itself. If you
load this AppMap without calling "normalize" then if you get an error, it
will be the original id as generated by the agent.

The loader probably does expect the ids to be monotonically increasing on a
thread. So if you have ids that go up and down, I could see it getting
confused.

Are you saying that a later-occurring event could have a lower id than an
early one? Event ids are normally assigned using a global counter that's
monotonically increasing, so events that are later in time always have a
higher numerical id.

- kevin

**Comment by lachrist** [comment]
Hi @kgilpin thanks for your answer. As you know, in the JS agent, I reorder events based on their current asynchronous id (basically the index of the current event loop execution). So events' id might indeed not be by monotonous. Btw because I do events reordering myself I always have but a single thread. If event monotonous indexing is a requirement shouldn't it be made explicit in the specification? I would argue that the best outcome would be to use event indexes only to match call/return and not attach any chronological meaning to them.

**Comment by lachrist** [comment]
Is there a way to disable the normalization phase and/or fetch to the real event id that is problematic?

**Comment by kgilpin** [comment]
Yes. Just call "buildAppMap" without calling "normalize"

**Comment by lachrist** [comment]
Ah ok, I'm just used to checkout appmaps directly within the vscode extension. I will take take a look at how to programmatically interact with `appmap-js`.

**Comment by kgilpin** [comment]
The code which loads AppMaps is assuming that when event ids are out of
order or missing, the agent has malformed the AppMap and it's attempting to
fix the problem.

If the agent omits event ids or parent ids, then the loader will assume
that the event order and thread ids are correct, and assign event ids and
parent ids accordingly.

If the agent assigns event ids and parent ids, then it assumes that these
are correct. We hadn't had a situation where the agent is assigning
non-increasing event ids on purpose.

We can either make a fix in the JS agent or change it in the loader's
"normalize" function. I can say that having non-increasing event ids is a
bit surprising.

**Comment by kgilpin** [comment]
Sure.

const { buildAppMap } = ***@***.***/models');

**Comment by lachrist** [comment]
Indeed, bypassing normalization makes it work. Also, I've re-indexed events and its works now (with the normalization). I still think the monotonous requirement on the indexes should be made explicit (or even better, relaxed!). Should I update the appmap spec?

**Comment by kgilpin** [comment]
Yes, please update the spec. I think it's reasonable to ask agents to do
this - and it's hard to deal with if they don't.
